Allow 'public' classes to have 'internal' required initializers

jrose-apple opened the pull request:
They're already not subclassable publicly, so it's okay for the initializer to not be available to cross-module clients, just like if it were non-'required'. This allows constructing a class instance chosen at runtime *within* the module without having to expose the existence of the constructor to everybody.

rdar://problem/22845087

jrose-apple commented:
@jckarter, @rjmccall, do you think this needs to go through Evolution? Or is it a simple consequence of the `open`/`public` distinction that we're just finally getting to now?

jrose-apple commented:
@swift-ci Please test

jckarter commented:
This seems reasonable to me. More generally, `required`/`designated`/`convenience` could all ideally remain an implementation detail for classes in a non-open class hierarchy.

jrose-apple commented:
Ah, you can still write *extensions* of a non-open class, so I'm not sure I want to go there yet. And if the initializer *is* public then it can still be initialized dynamically across module boundaries.

jrose-apple committed:
Allow 'public' classes to have 'internal' required initializers

They're already not subclassable publicly, so it's okay for the
initializer to not be available to cross-module clients, just like if
it were non-'required'. This allows constructing a class instance
chosen at runtime /within/ the module without having to expose the
existence of the constructor to everybody.

rdar://problem/22845087

jrose-apple commented:
@swift-ci Please smoke test
